pkg/token/serviceprincipaltoken_test.go

```diff
@@ -0,0 +1,232 @@
+	rec.SetRealTransport(&http.Transport{
+		TLSNextProto: make(map[string]func(authority string, c *tls.Conn) http.RoundTripper),
+		// #nosec G402: PAV2 service only supports TLS12
+		TLSClientConfig: &tls.Config{
+			Renegotiation: tls.RenegotiateFreelyAsClient,
+			MinVersion:    tls.VersionTLS12,
+			MaxVersion:    tls.VersionTLS12,
+		},
+	})
+
+	hook := func(i *cassette.Interaction) error {
+		// Delete sensitive content
+		delete(i.Response.Headers, "Set-Cookie")
+		delete(i.Response.Headers, "X-Ms-Request-Id")
+		if i.Request.Form["client_id"] != nil {
+			i.Request.Form["client_id"] = []string{redactionToken}
+		}
+		if i.Request.Form["client_secret"] != nil && i.Request.Form["client_secret"][0] != badSecret {
+			i.Request.Form["client_secret"] = []string{redactionToken}
+		}
+		if i.Request.Form["client_assertion"] != nil {
+			i.Request.Form["client_assertion"] = []string{redactionToken}
+		}
+		if i.Request.Form["scope"] != nil {
+			i.Request.Form["scope"] = []string{redactionToken + "/.default openid offline_access profile"}
+		}
+		i.Request.URL = strings.ReplaceAll(i.Request.URL, os.Getenv(tenantID), tenantID)
+		i.Response.Body = strings.ReplaceAll(i.Response.Body, os.Getenv(tenantID), tenantID)
+
+		if strings.Contains(i.Request.Body, "client_secret") {
+			i.Request.Body = `client_id=[REDACTED]&client_secret=[REDACTED]&grant_type=client_credentials&scope=[REDACTED]%2F.default+openid+offline_access+profile`
+		}
+
+		if strings.Contains(i.Request.Body, "client_assertion") {
+			i.Request.Body = `client_assertion=[REDACTED]&client_assertion_type=urn%3Aietf%3Aparams%3Aoauth%3Aclient-assertion-type%3Ajwt-bearer&client_id=[REDACTED]&client_info=1&grant_type=client_credentials&scope=[REDACTED]%2F.default+openid+offline_access+profile`
+		}
+
+		if strings.Contains(i.Response.Body, "access_token") {
+			i.Response.Body = `{"token_type":"Bearer","expires_in":86399,"ext_expires_in":86399,"access_token":"[REDACTED]"}`
+		}
+		return nil
+	}
+	rec.AddHook(hook, recorder.BeforeSaveHook)
+
+	rec.SetMatcher(customMatcher)
+	rec.SetReplayableInteractions(true)
+
+	return rec, rec.GetDefaultClient()
+}
+
+func customMatcher(r *http.Request, i cassette.Request) bool {
+	id := os.Getenv(tenantID)
+	if id == "" {
+		id = "00000000-0000-0000-0000-000000000000"
+	}
+	switch os.Getenv(vcrMode) {
+	case vcrModeRecordOnly:
+	default:
+		r.URL.Path = strings.ReplaceAll(r.URL.Path, id, tenantID)
+	}
+	return cassette.DefaultMatcher(r, i)
+}
+
+// Get go-vcr record mode from enviroment variable
+func getVCRMode() recorder.Mode {
+	switch os.Getenv(vcrMode) {
+	case vcrModeRecordOnly:
+		return recorder.ModeRecordOnly
+	default:
+		return recorder.ModeReplayOnly
+	}
+}
+
+func TestServicePrincipalLoginVCR(t *testing.T) {
+	testCase := []struct {
+		cassetteName  string
+		description   string
+		p             *servicePrincipalToken
+		expectedError error
+	}{
+		{
+			cassetteName: "BadSecretVCR",
+			description:  "Test using incorrect secret value",
```

Comment:
you can save one property by using one of them as cassette name. For instance
```go
{
name: "Test using incorrect secret value",
...
}

t.Run(tc.name, ...) {
vcrRecorder, httpClient := getVCRHttpClient(t.Name())
```
}

Reply:
Fixed

Reply:
Fixed